Refactor URL validation CLI script to make parts reusable

johnwatkins0 opened the pull request:
## Summary

This refactors the URL validation CLI script. The functions that generate URLs to validate are in the new class `ScannableURLProvider`, and the functions that actually validate URLs are in the new class `URLValidationProvider`.

This fixes #5307 and is a precursor to several open issues, including anything that involves validating URLs without the loopback request -- e.g. via cron tasks or in client-side requests via REST. In the closed PR #5228, where most of this code comes from, I attempted to refactor the CLI script _and_ create REST wrappers _and_ create a Cron service. In hindsight, that was doing too much, and I think it would be better to get this change reviewed and merged before doing any work that depends on it.

While most of the code in this PR is from #5228, it is in turn mostly from the CLI script. Even the PHPUnit tests were mostly copied into new classes. For that reason, the new classes are still architected with the CLI script in mind, so I believe they may evolve more as we work on the features that depend on them. 

### Testing

As mentioned above, most of the unit tests here are slightly modified versions of the tests covering the CLI scripts, so their passing is a big indicator that nothing has broken. For QA steps, see #5307.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Refactor URL validation CLI script

johnwatkins0 committed:
URL validation: return WP_Error if validity is an error

johnwatkins0 committed:
Use with_lock where method is called instead of inside method

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+ *
+ * @since 2.1
+ */
+final class ValidationURLProvider {
```
To avoid confusion with the `UrlValidationProvider`, I suggest renaming this to `ScannableUrlProvider`.

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * @param integer $limit_per_type The maximum number of URLs to validate for each type.
+	 * @param array   $include_conditionals An allowlist of conditionals to use for validation.
+	 * @param boolean $include_unsupported Whether to include URLs that don't support AMP.
```
```suggestion
	 * @param integer $limit_per_type       The maximum number of URLs to validate for each type.
	 * @param array   $include_conditionals An allowlist of conditionals to use for validation.
	 * @param boolean $include_unsupported  Whether to include URLs that don't support AMP.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	}
+
+	/**
+	 * Provides the array of URLs to check. Each URL is an array with two elements, with the URL at index 0 and the type at index 1.
```
This long line can be split into short and long description.
```suggestion
	 * Provides the array of URLs to check.
	 *
	 * Each URL is an array with two elements, with the URL at index 0 and the type at index 1.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	/**
+	 * Provides the array of URLs to check. Each URL is an array with two elements, with the URL at index 0 and the type at index 1.
+	 *
+	 * @param int|null $offset The number of URLs to offset by, where applicable.
```
```suggestion
	 * @param int|null $offset Optional. The number of URLs to offset by, where applicable. Defaults to 0.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 * @return bool Whether the template is supported.
+	 */
+	public function is_template_supported( $template ) {
+		// If the --include argument is present in the WP-CLI command, this template conditional must be present in it.
```
This class is supposed to be independent from the WP-CLI command and shouldn't have knowledge about it, so this comment seems off.
```suggestion
		// If we received an allowlist of conditionals, this template conditional must be present in it.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * Usually, this class will query all of the templates that don't have AMP disabled. This allows inclusion based on only these conditionals.
+	 *
+	 * @var array
```
```suggestion
	 * @var string[]
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+
+		/*
+		 * Handle the argument and flag passed to the command: --include and --force.
+		 * If the self::INCLUDE_ARGUMENT is present, force crawling or URLs.
```
```suggestion
		 * If the self::INCLUDE_ARGUMENT is present, force crawling of URLs.
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+			],
+			$this->assoc_args
+		);
+	}
```
This is not how WP-CLI associative arguments are supposed to be handled. Instead of managing defaults separately here and then doing `isset()` checks, you'd rather use `WP_CLI\Utils::get_flag_value( $assoc_args, <flag_name>, <default_value>)`.

This is important because the `get_flag_value()` helper ensures all additional logic keeps working, like the ability to negate any boolean flag by prefixing it with `no-`, like `--no-force`.
```suggestion
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+
+		$include_conditionals      = [];
+		$force_crawl_urls          = false;
+		$limit_type_validate_count = (int) $assoc_args[ self::LIMIT_URLS_ARGUMENT ];
```
This should use the `WP_CLI\Utils::get_flag_value()` helper to ensure all additional flag logic keeps working (see comment above).
```suggestion
		$limit_type_validate_count = Utils::get_flag_value( $assoc_args, self::LIMIT_URLS_ARGUMENT, 100 );
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+		}
+
+		$assoc_args = $this->get_assoc_args();
+
```
This should not be needed.
```suggestion
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+		$assoc_args = $this->get_assoc_args();
+
+		$include_conditionals      = [];
+		$force_crawl_urls          = false;
```
This can immediately be set to the force flag and adapted as needed later.
```suggestion
		$force_crawl_urls          = Utils::get_flag_value( $assoc_args, self::FLAG_NAME_FORCE_VALIDATION, false );
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+
+		$assoc_args = $this->get_assoc_args();
+
+		$include_conditionals      = [];
```
This can immediately be set to the includes flag and adapted as needed later.
```suggestion
		$include_conditionals      = Utils::get_flag_value( $assoc_args, self::INCLUDE_ARGUMENT, [] );
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+		 * If the self::INCLUDE_ARGUMENT is present, force crawling or URLs.
+		 * The WP-CLI command should indicate which templates are crawled, not the /wp-admin options.
+		 */
+		if ( ! empty( $assoc_args[ self::INCLUDE_ARGUMENT ] ) ) {
```
```suggestion
		if ( ! empty( $include_conditionals ) ) {
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +194,129 @@ public function run( $args, $assoc_args ) {
+			$force_crawl_urls     = true;
+		} elseif ( isset( $assoc_args[ self::FLAG_NAME_FORCE_VALIDATION ] ) ) {
+			$force_crawl_urls = true;
+		}
```
This is already taken care of above.
```suggestion
		}
```

schlessera commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,216 @@
+	 *
+	 * @var int
+	 */
+	public $total_errors = 0;
```
Why are these 4 properties public? A public property generally defeats the purpose of using an object in the first place, as you lose control over the state of the data and thus can never rely on anything about it.

These should be `private`, with separate getter methods (but only if they are actually needed).

schlessera commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,216 @@
+	 * Locks validation.
+	 */
+	private function lock() {
+		set_transient( self::LOCK_TRANSIENT, 'locked', $this->get_lock_timeout() );
```
A transient is not a reliable lock mechanism, as there's no guarantee it will even be stored the next time you want to check it. Transients can only be reliably used for volatile data that will be recreated whenever needed.
A lock is not volatile data, as this would probably mean certain data corruption.
The lock mechanism should rather use a non-autoloaded option, I think.

schlessera commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,216 @@
+		}
+
+		$this->lock();
+		$result = call_user_func( $callback );
```
This should just execute the callable (which is also much faster):
```suggestion
		$result = $callback();
```

schlessera commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,216 @@
+			return $validity;
+		} elseif ( $validity && isset( $validity['results'] ) ) {
+			$this->update_state_from_validity( $validity, $type );
+		}
```
No need for an else, as the `return` already implies that:
```suggestion
		}

		if ( $validity && isset( $validity['results'] ) ) {
			$this->update_state_from_validity( $validity, $type );
		}
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * @var bool
+	 */
+	public $include_unsupported;
```
Again, `public` properties should be avoided and turned into private properties with getters (if needed).

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	}
+
+	/**
+	 * Gets the posts IDs that support AMP.
```
```suggestion
	 * Gets the post IDs that support AMP.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * By default, this only gets the post IDs if they support AMP.
+	 * This means that 'Posts' isn't deselected in 'AMP Settings' > 'Supported Templates'.
+	 * And 'Enable AMP' isn't unchecked in the post's editor.
```
```suggestion
	 * This means that 'Posts' isn't deselected in 'AMP Settings' > 'Supported Templates'
	 * and 'Enable AMP' isn't unchecked in the post's editor.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 * This means that 'Posts' isn't deselected in 'AMP Settings' > 'Supported Templates'.
+	 * And 'Enable AMP' isn't unchecked in the post's editor.
+	 *
+	 * @param array $ids The post IDs to check for AMP support.
```
```suggestion
	 * @param int[] $ids The post IDs to check for AMP support.
```

I just noticed that the typing information for `post_supports_amp()` and its deprecated cousin is off, as they accept `WP_Post[]|int[]`.

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * @param int|string $offset The offset for the URL to query for, should be an int if passing an argument.
+	 * @param int|string $number The total number to query for, should be an int if passing an argument.
+	 * @return array The author page URLs, or an empty array.
```
```suggestion
	 * @return string[] The author page URLs, or an empty array.
```

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 *
+	 * This only gets the term IDs if they support AMP.
+	 * If their taxonomy is unchecked in 'AMP Settings' > 'Supported Templates,' this does not return them.
+	 * For example, if 'Categories' is unchecked.
```
This comment seems off, as this method returns a `bool`.

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+final class URLValidationProviderTest extends WP_UnitTestCase {
+	use PrivateAccess, AssertContainsCompatibility;
+
+	/**
```
Properties should be explicitly declared and not dynamically created.
```suggestion
	/**
	 * Validation provider instance to use.
	 *
	 * @var URLValidationProvider
	 */
	private $validation_provider;

	/**
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+	 */
+	public function test_get_url_validation() {
+		$single_post_permalink = get_permalink( self::factory()->post->create() );
+		$this->call_private_method( $this->validation_provider, 'get_url_validation', [ $single_post_permalink, 'post' ] );
```
This is not a private method.
```suggestion
		$this->validation_provider->get_url_validation( $single_post_permalink, 'post' );
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+		for ( $i = 0; $i < $number_of_posts; $i++ ) {
+			$permalink         = get_permalink( self::factory()->post->create() );
+			$post_permalinks[] = $permalink;
+			$this->call_private_method( $this->validation_provider, 'get_url_validation', [ $permalink, 'post' ] );
```
```suggestion
			$this->validation_provider->get_url_validation( $permalink, 'post' );
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+
+/** @coversDefaultClass URLValidationProvider */
+final class URLValidationProviderTest extends WP_UnitTestCase {
+	use PrivateAccess, AssertContainsCompatibility;
```
Private access is not needed.
```suggestion
	use AssertContainsCompatibility;
```

schlessera replied:
Actually, both of these traits are not needed, as you're also not using `assertStringContains()`.
```suggestion
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+
+		$expected_result = 'EXPECTED RESULT';
+		$result          = $this->validation_provider->with_lock(
+			function() use ( $expected_result ) {
```
The `function` term is not the name of the function. Instead, there is no name, and that absence is what has the parentheses attached, not the `function` keyword.
It's basically: `function <no-name>()`, so the parentheses shouldn't be attached to `function`. Otherwise, it looks like you could execute `function`...
```suggestion
			function () use ( $expected_result ) {
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+		// Test with_lock with no return value.
+		$this->assertNull(
+			$this->validation_provider->with_lock(
+				function() {
```
```suggestion
				function () {
```

schlessera commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -222,15 +138,26 @@ public function run( $args, $assoc_args ) {
-		$this->crawl_site();
+
+		$result = $validation_provider->with_lock(
+			function() {
```
```suggestion
			function () {
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+	 */
+	public function setUp() {
+		parent::setUp();
+		$this->validation_provider = new URLValidationProvider( 100 );
```
`URLValidationProvider` does not have a constructor, so the `100` does not do anything here.
```suggestion
		$this->validation_provider = new URLValidationProvider();
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+	public function test_get_url_validation() {
+		$single_post_permalink = get_permalink( self::factory()->post->create() );
+		$this->call_private_method( $this->validation_provider, 'get_url_validation', [ $single_post_permalink, 'post' ] );
+		$this->assertTrue( in_array( $single_post_permalink, ValidationRequestMocking::get_validated_urls(), true ) );
```
```suggestion
		$this->assertContains( $single_post_permalink, ValidationRequestMocking::get_validated_urls() );
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+			$this->validation_provider->with_lock(
+				function() {
+					$this->assertEquals( 'locked', get_transient( URLValidationProvider::LOCK_TRANSIENT ) );
+					return;
```
This `return;` doesn't do anything.
```suggestion
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+			$post_ids[0],
+			$terms_for_current_taxonomy,
+			$taxonomy
+		);
```
Maybe assert this call worked as intended?
```suggestion
		$result = wp_set_post_terms(
			$post_ids[0],
			$terms_for_current_taxonomy,
			$taxonomy
		);
		$this->assertFalse( is_wp_error( $result ) );
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+			$taxonomy
+		);
+
+		$this->assertEquals( $expected_url_count, count( $this->validation_url_provider->get_urls() ) );
```
```suggestion
		$this->assertCount( $expected_url_count, $this->validation_url_provider->get_urls() );
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+
+		$number_original_urls = 4;
+
+		$this->assertEquals( $number_original_urls, count( $this->validation_url_provider->get_urls() ) );
```
```suggestion
		$this->assertCount( $number_original_urls, $this->validation_url_provider->get_urls() );
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+		 * And ensure that the tested method finds a URL for all of them.
+		 */
+		$expected_url_count = $number_new_posts + $number_original_urls + 1;
+		$this->assertEquals( $expected_url_count, count( $this->validation_url_provider->get_urls() ) );
```
```suggestion
		$this->assertCount( $expected_url_count, $this->validation_url_provider->get_urls() );
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+	}
+
+	/**
+	 * Test retreival of urls.
```
```suggestion
	 * Test retrieval of urls.
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+final class ValidationURLProviderTest extends WP_UnitTestCase {
+	use PrivateAccess, AssertContainsCompatibility;
+
+	/**
```
```suggestion
	/**
	 * Validation URL provider instance to use.
	 *
	 * @var ValidationURLProvider
	 */

	/**
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+	 */
+	public function test_count_urls_to_validate() {
+		$this->validation_url_provider = new ValidationURLProvider();
+
```
This is already done in `setUp()`.
```suggestion
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+		$this->assertEquals( [], $this->call_private_method( $this->validation_url_provider, 'get_posts_that_support_amp', [ $ids ] ) );
+
+		/*
+		 * If is_singular is in the WP-CLI argument, it should allow return these posts as being AMP-enabled.
```
```suggestion
		 * If is_singular is in the WP-CLI argument, it should return these posts as being AMP-enabled.
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+
+		AMP_Options_Manager::update_option( Option::SUPPORTED_TEMPLATES, [ $author_conditional ] );
+		AMP_Options_Manager::update_option( Option::ALL_TEMPLATES_SUPPORTED, false );
+		$this->assertTrue( $this->call_private_method( $this->validation_url_provider, 'is_template_supported', [ $author_conditional ] ) );
```
`ValidationURLProvider::is_template_supported()` is currently marked as `public`, not `private`, so the `$this->call_private_method()` would not be needed here.

However, I wonder whether the public access is intentional or not...?

schlessera commented on src/Validation/ValidationURLProvider.php:
```diff
@@ -0,0 +1,307 @@
+	 * @param string $template The template to check.
+	 * @return bool Whether the template is supported.
+	 */
+	public function is_template_supported( $template ) {
```
I don't think that this is meant to be `public`, as it is only internally used by `ValidationURLProvider` right now.
```suggestion
	private function is_template_supported( $template ) {
```

schlessera commented on tests/php/src/Validation/ValidationURLProviderTest.php:
```diff
@@ -0,0 +1,386 @@
+				self::factory()->post->create(),
+				$terms_for_current_taxonomy,
+				$taxonomy
+			);
```
Maybe assert this call worked as intended?
```suggestion
			$result = wp_set_post_terms(
 				self::factory()->post->create(),
 				$terms_for_current_taxonomy,
 				$taxonomy
 			);
 			$this->assertFalse( is_wp_error( $result ) );
```

schlessera commented on tests/php/test-class-amp-cli-validation-command.php:
```diff
@@ -407,120 +59,24 @@ public function test_crawl_site() {
-				'message' => 'ok',
-			],
-		];
+		$this->assertTrue( in_array( home_url( '/' ), ValidationRequestMocking::get_validated_urls(), true ) );
```
```suggestion
		$this->assertContains( home_url('/'), ValidationRequestMocking::get_validated_urls() );
```

schlessera commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,96 @@
+namespace AmpProject\AmpWP\Tests\Validation;
+
+use AmpProject\AmpWP\Tests\Helpers\AssertContainsCompatibility;
+use AmpProject\AmpWP\Tests\Helpers\PrivateAccess;
```
```suggestion
```

schlessera commented on tests/php/test-class-amp-cli-validation-command.php:
```diff
@@ -8,11 +8,14 @@
 use AmpProject\AmpWP\Option;
 use AmpProject\AmpWP\Tests\Helpers\AssertContainsCompatibility;
```
```suggestion
```

This is not needed.

You need to remove the `use` statement for this as well below here, I can't reach it with the PR review tools.

schlessera reviewed (changes requested):
Great improvement. I think the naming should be revised between `ValidationURLProvider` and `URLValidationProvider`, as these are almost guaranteed to confuse people. I think you've even confused yourself in the tests, as you've used the constructor from one on the other... ;)

There are lots of suggestions in the review, but a large part of it is just nit-picking.

johnwatkins0 committed:
Minor updates primarily in formatting and comments

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

johnwatkins0 committed:
Use get_flag_value in CLI script to handle assoc args

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

johnwatkins0 committed:
PHPCS fixes

johnwatkins0 committed:
Merge branch 'feature/refactor-url-validation' of https://github.com/ampproject/amp-wp into feature/refactor-url-validation

johnwatkins0 committed:
Fix CLI script imports

johnwatkins0 committed:
Use option instead of transient for locking

johnwatkins0 committed:
Rename ValidationURLProvider

johnwatkins0 committed:
Make class properties private

johnwatkins0 committed:
Trim outdated comment content

johnwatkins0 committed:
Remove unneeded use

johnwatkins0 committed:
Remove test code

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/refactor-url-validation

schlessera committed:
Try importing WP_CLI\Utils

schlessera commented:
@googlebot I consent.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=5306) <br/>All committers have signed the CLA.

johnwatkins0 commented:
@googlebot I consent.

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/refactor-url-validation

johnwatkins0 commented:
This PR is currently blocked by this in the PHPUnit tests:

```
Error: Call to undefined function WP_CLI\Utils\get_flag_value()
```

That WP_CLI util function does exist in my local environment, so investigation is needed into why the test environment doesn't have it.

johnwatkins0 committed:
Include cli utils in autoload-dev

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/refactor-url-validation

johnwatkins0 committed:
Remove cli utils from autoload-dev

johnwatkins0 committed:
Fix with_lock unit tests

johnwatkins0 commented:
@schlessera Thanks for your feedback and your help with the unit tests. All comments have been addressed, and I've got the tests passing. Re-requesting review.

pierlon commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -157,47 +109,12 @@ final class AMP_CLI_Validation_Command {
-		$number_urls_to_crawl = $this->count_urls_to_validate();
+		$number_urls_to_crawl = count( $validation_url_provider->get_urls() );
 		if ( ! $number_urls_to_crawl ) {
 			if ( ! empty( $this->include_conditionals ) ) {
```
`$this->include_conditionals` doesn't seem to belong to this class anymore.

johnwatkins0 replied:
Fixed in 58bacd089

pierlon commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -222,15 +139,26 @@ public function run( $args, $assoc_args ) {
 
 		$table_validation_by_type = [];
-		foreach ( $this->validity_by_type as $type_name => $validity ) {
+		foreach ( $validation_provider->validity_by_type as $type_name => $validity ) {
```
I'd recommend using a getter to retrieve `$validation_provider->validity_by_type` and make `$validity_by_type` in `URLValidationProvider` non-public.

johnwatkins0 replied:
Done in 050b1e74c

pierlon commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -267,6 +195,108 @@ public function run( $args, $assoc_args ) {
+		foreach ( $validation_url_provider->get_urls() as $url ) {
+			$validity = $validation_provider->get_url_validation( $url['url'], $url['type'], URLValidationProvider::FLAG_FORCE_REVALIDATE );
+
+			if ( $this->wp_cli_progress ) {
```
Wouldn't `$this->wp_cli_progress` already be set?

johnwatkins0 replied:
Normally yes, but this method is directly tested in PHPUnit, so `wp_cli_progress` is not set in that case. I could add a method to create `wp_cli_progress` on the fly when needed, but that might not be worth the additional complexity. Let me know if you disagree.

pierlon commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -0,0 +1,305 @@
+	private $limit_per_type;
+
+	/**
+	 * Class construÎctor.
```
```suggestion
	 * Class constructor.
```

pierlon commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -0,0 +1,305 @@
+			get_taxonomies( [ 'public' => true ] ),
+			[ $this, 'does_taxonomy_support_amp' ]
+		);
+		$public_post_types      = get_post_types( [ 'public' => true ], 'names' );
```
The `$output` parameter for `get_post_types()` is `'names'` by default.

```suggestion
		$public_post_types      = get_post_types( [ 'public' => true ] );
```

pierlon commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -0,0 +1,305 @@
+	 * @param int|null $number The number of posts to query for (optional).
+	 * @return int[]   $post_ids The post IDs in an array.
+	 */
+	private function get_posts_by_type( $post_type, $offset = null, $number = null ) {
```
Any reason why the `$number` parameter for `get_posts_by_type()`, `get_posts_by_type()` and `get_taxonomy_links()` have different different default values? I'd recommend standardizing them, if possible.

pierlon commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,248 @@
+	 * @return boolean
+	 */
+	public function is_locked() {
+		$lock_time = intval( get_option( self::LOCK_KEY, 0 ) );
```
This is a matter of personal preference, but I'd say typecasting to `int` is easier to read. Typecasting is also much faster than using `intval()`, but it seems it's you only notice the difference [under heavy workload](https://stackoverflow.com/a/35949025).

/cc @schlessera @westonruter for any additional insight.

```suggestion
		$lock_time = (int) get_option( self::LOCK_KEY, 0 );
```

pierlon commented on tests/php/src/Validation/URLValidationProviderTest.php:
```diff
@@ -0,0 +1,94 @@
+use AmpProject\AmpWP\Validation\URLValidationProvider;
+use WP_UnitTestCase;
+
+/** @coversDefaultClass URLValidationProvider */
```
```suggestion
/** @coversDefaultClass \AmpProject\AmpWP\Validation\URLValidationProvider */
```

pierlon commented on tests/php/src/Validation/ScannableURLProviderTest.php:
```diff
@@ -0,0 +1,392 @@
+use WP_Query;
+use WP_UnitTestCase;
+
+/** @coversDefaultClass ScannableURLProvider */
```
```suggestion
/** @coversDefaultClass \AmpProject\AmpWP\Validation\ScannableURLProvider */
```

pierlon commented on tests/php/src/Validation/ScannableURLProviderTest.php:
```diff
@@ -0,0 +1,392 @@
+	 * Validation URL provider instance to use.
+	 *
+	 * @var ScannableURLProvider
+	 */
```
Looks like you forgot to declare `$validation_url_provider` as a property.

pierlon commented on tests/php/src/Validation/ScannableURLProviderTest.php:
```diff
@@ -0,0 +1,392 @@
+	 */
+	public function test_get_search_page() {
+		// Normally, this should return a string, unless the user has opted out of the search template.
+		$this->assertTrue( is_string( $this->call_private_method( $this->validation_url_provider, 'get_search_page' ) ) );
```
The `assertIsString()` assertion seems more appropriate.

```suggestion
		$this->assertIsString( $this->call_private_method( $this->validation_url_provider, 'get_search_page' ) );
```

pierlon commented on tests/php/src/Validation/ScannableURLProviderTest.php:
```diff
@@ -0,0 +1,392 @@
+
+		// If $include_conditionals has is_search, this should return a URL.
+		$this->set_private_property( $this->validation_url_provider, 'include_conditionals', [ 'is_search' ] );
+		$this->assertTrue( is_string( $this->call_private_method( $this->validation_url_provider, 'get_search_page' ) ) );
```
```suggestion
		$this->assertIsString( $this->call_private_method( $this->validation_url_provider, 'get_search_page' ) );
```

johnwatkins0 committed:
Apply suggestions from code review

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Replace obsolete class property with get_flag_value call

johnwatkins0 committed:
Make validity_by_type class property private and add getter method

johnwatkins0 committed:
ScannableURLProvider: return limit_per_type # of taxonomy links by default

johnwatkins0 committed:
Add omitted property

johnwatkins0 committed:
Undo use of assertIsString

johnwatkins0 committed:
PHPStan ignore a line

johnwatkins0 committed:
Update lock file

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/refactor-url-validation

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/refactor-url-validation

johnwatkins0 committed:
Simplify test assertion

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

johnwatkins0 committed:
Improve coverage with more @covers comments

johnwatkins0 committed:
Update covers comment style

johnwatkins0 committed:
PHPCS fix

johnwatkins0 commented:
@pierlon @westonruter @schlessera I've addressed all feedback and gotten the new GHA workflow to pass. This is ready for another round of review. 

My not-quite-finished PR adding the validation cron events (#5515) depends on this.

pierlon commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -60,76 +63,25 @@ final class AMP_CLI_Validation_Command {
+	 * @var ScannableURLProvider
 	 */
-	public $force_crawl_urls = false;
+	private $validation_url_provider;
```
I wonder if `$validation_url_provider` and `$validation_provider` are misnomers.

I would expect `$validation_url_provider`to be named something like `$scannable_url_provider` since it is an instance of `ScannableURLProvider`. Likewise, I would expect `$validation_provider`to be named something like `$url_validation_provider` since it is an instance of `URLValidationProvider`.

johnwatkins0 replied:
Yes, you're right. This was a product of shifting class names. Updated in 4c0238d76

pierlon commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -157,49 +109,15 @@ final class AMP_CLI_Validation_Command {
+		$number_urls_to_crawl = count( $validation_url_provider->get_urls() );
 		if ( ! $number_urls_to_crawl ) {
-			if ( ! empty( $this->include_conditionals ) ) {
+			/** @phpstan-ignore-next-line */
```
This shouldn't be needed. As I understand it you would have to stub the `Utils\get_flag_value()` method, which can be done by:

```diff
diff --git a/includes/cli/class-amp-cli-validation-command.php b/includes/cli/class-amp-cli-validation-command.php
--- a/includes/cli/class-amp-cli-validation-command.php	(revision 6c449c04cf81c7c1b7228e07ed881922a68802a7)
+++ b/includes/cli/class-amp-cli-validation-command.php	(date 1605052727456)
@@ -116,7 +116,6 @@
 
 		$number_urls_to_crawl = count( $validation_url_provider->get_urls() );
 		if ( ! $number_urls_to_crawl ) {
-			/** @phpstan-ignore-next-line */
 			if ( ! empty( Utils\get_flag_value( $this->assoc_args, self::INCLUDE_ARGUMENT, [] ) ) ) {
 				WP_CLI::error(
 					sprintf(

diff --git a/tests/php/static-analysis-stubs/wp-cli.php b/tests/php/static-analysis-stubs/wp-cli.php
--- a/tests/php/static-analysis-stubs/wp-cli.php	(revision 6c449c04cf81c7c1b7228e07ed881922a68802a7)
+++ b/tests/php/static-analysis-stubs/wp-cli.php	(date 1605053104533)
@@ -83,4 +83,23 @@
     function make_progress_bar($message, $count, $interval = 100)
     {
     }
+
+	/**
+	 * Return the flag value or, if it's not set, the $default value.
+	 *
+	 * Because flags can be negated (e.g. --no-quiet to negate --quiet), this
+	 * function provides a safer alternative to using
+	 * `isset( $assoc_args['quiet'] )` or similar.
+	 *
+	 * @access public
+	 * @category Input
+	 *
+	 * @param array  $assoc_args Arguments array.
+	 * @param string $flag       Flag to get the value.
+	 * @param mixed  $default    Default value for the flag. Default: NULL.
+	 * @return mixed
+	 */
+	function get_flag_value( $assoc_args, $flag, $default = null )
+	{
+	}
 }
```

pierlon replied:
You may need to clear the `phpstan` result cache (`vendor/bin/phpstan clear-result-cache`) before running static analysis again.

johnwatkins0 replied:
That seemed to work. updated in 6f42fd3

pierlon commented on tests/php/src/Validation/ScannableURLProviderTest.php:
```diff
@@ -0,0 +1,398 @@
+		add_filter( 'pre_http_request', [ ValidationRequestMocking::class, 'get_validate_response' ] );
+	}
+
+	/** @covers ::__construct */
```
We don't want to commit perjury and upset the senate now :wink: 

```suggestion
	/** @covers ::__construct() */
```

pierlon reviewed (changes requested):
Some misc. changes to be made and then all should be well.

johnwatkins0 committed:
Rename class properties to match the classes they're instances of

johnwatkins0 committed:
Add stub for wp cli function to avoid phpstan ignore comment

johnwatkins0 committed:
Add parentheses to function in covers tag

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 commented:
Thanks, @pierlon. Updated and re-requesting review.

pierlon reviewed (approved):
Well done!

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/refactor-url-validation

* 'develop' of github.com:ampproject/amp-wp: (33 commits)
  Ensure global post is set for test_decorate_shortcode_and_filter_source
  Extract to plugins dir when unzipping PWA plugin
  Add tests for new markup
  Use fixed layout for amp-story-360 example
  Undo intrinsic layout for video since WP core uses responsive
  Use intrinsic layout of video
  Update amphtml spec to 2011140244000
  Update stable tag and tested up to
  Bump stable tag to 2.0.6; update docs
  Bump autoprefixer from 10.0.1 to 10.0.2
  Bump babel-loader from 8.1.0 to 8.2.1
  Setup composer via the shivammathur/setup-php action
  Bump core-js from 3.6.5 to 3.7.0
  Bump copy-webpack-plugin from 6.3.0 to 6.3.1
  Remove deprecated option "--no-suggest"
  Update composer.lock
  Bump eslint-plugin-jest from 24.1.0 to 24.1.3
  Bump eslint from 7.12.1 to 7.13.0
  Bump @wordpress/block-editor from 5.1.2 to 5.1.3
  Bump mini-css-extract-plugin from 1.2.1 to 1.3.1
  ...

westonruter commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -222,15 +139,26 @@ public function run( $args, $assoc_args ) {
+		if ( is_wp_error( $result ) ) {
+			WP_CLI::error( 'The site cannot be crawled at this time because validation is running in another process.' );
+			return;
+		}
```
Question: Why is a lock necessary? What problems will occur if validation is happening in parallel?

johnwatkins0 replied:
Because the `validate_urls` method ultimately leads to `AMP_Validation_Manager::validate_url`, which is an expensive function using `wp_remote_get`. In theory, without the lock, if someone runs the CLI command (or even multiple instances of the command) there could be multiple processes running a string of `wp_remote_get` calls at the same time. I suspect some hosting providers wouldn't like that. Do you think the lock is overcautious? Removing it would certainly simplify things.

johnwatkins0 replied:
Forgot to note here, the lock is looking ahead to having the cron task in place, as well as any URL validation taking place within the plugins settings screen.

westonruter replied:
OK, I'm still not sure on the need for locking but we can revisit that later. It may be preferable for any new invocation of the validation provider to abort any existing validation tasks. Then you wouldn't have the experience of validation being performed in cron, and then being blocked when attempting to initiate validation from the admin screen. 

Nevertheless, validation happening concurrently generally should not have negative consequences other than additional server load. Even if the same URL is being validated at the exact same time, what would go wrong?

johnwatkins0 replied:
I'll be following up shortly with the cron PR that uses the lock more. That will be a good time to revisit. My fundamental assumption is that hosting platforms like VIP might see multiple background processes running `wp_remote_get` and not liking that too much, but I might be overcomplicating things because I'm sure this is not the only plugin that does things like that.

westonruter commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -157,49 +109,14 @@ final class AMP_CLI_Validation_Command {
+		$url_validation_provider = $this->get_validation_provider();
 
-		$number_urls_to_crawl = $this->count_urls_to_validate();
+		$number_urls_to_crawl = count( $scannable_url_provider->get_urls() );
```
Should the obtained URLs be stored in a variable and check the `count()`? The URLs could then be passed down to the `validate_urls()` method. Otherwise, as it stands right now the `get_urls()` method is being called twice, with the same data (or not) being queried twice.

johnwatkins0 replied:
Done in de33111e05c1d05ff4ebcb5adda367bf7a328568

westonruter commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,257 @@
+		 *
+		 * @param int $timeout Time in seconds. Default 300 seconds.
+		 */
+		return apply_filters( 'amp_validation_lock_timeout', 5 * MINUTE_IN_SECONDS );
```
What is the use case for the filter? Should this be `@internal`?

johnwatkins0 replied:
Good question. I don't actually see an immediate use case for this filter, so I removed it in favor of a class constant in 13917b3b4

westonruter commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,257 @@
+					AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_ACK_ACCEPTED_STATUS !== $validation_status['term_status']
+					&&
+					AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_NEW_ACCEPTED_STATUS !== $validation_status['term_status']
+					);
```
```suggestion
					return (
						AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_ACK_ACCEPTED_STATUS !== $validation_status['term_status']
						&&
						AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_NEW_ACCEPTED_STATUS !== $validation_status['term_status']
					);
```

westonruter commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,257 @@
+				'total' => 0,
+			];
+		}
+			$this->validity_by_type[ $type ]['total']++;
```
Strange that PHPCS did not catch this:

```suggestion
		$this->validity_by_type[ $type ]['total']++;
```

westonruter commented on src/Validation/URLValidationProvider.php:
```diff
@@ -0,0 +1,257 @@
+	 *
+	 * @param string $url  The URL to validate.
+	 * @param string $type The type of template, post, or taxonomy.
+	 * @param string $flag Flag determining whether the URL should be revalidated.
```
A flag makes me think it is going to be a bitmask. Maybe `$mode` would be a better name?

westonruter replied:
Or actually, passing an arg for `FLAG_FORCE_REVALIDATE` vs `FLAG_NO_REVALIDATE` behavior may not be the most ergonomic. What if there were other methods for the other behaviors that wrap this one? If the flag is empty, then what is the expected behavior? Validate if it doesn't already exist? I think I see that now, but I had to look at it awhile to understand how it works.

johnwatkins0 replied:
Revised this in d87ad9d978a63df37fa7007c45f6ad1a6047311b. My solution for now is to simply use a third argument for `$force_revalidate`. The reason for this is I found `FLAG_NO_REVALIDATE` is not being used on this branch at all. The no-revalidate use case came up in one of the subsequent features that depend on this branch, and if needed I will revisit when I get back to that feature.

westonruter replied:
Sounds great.

westonruter commented on tests/php/src/Helpers/ValidationRequestMocking.php:
```diff
@@ -0,0 +1,63 @@
+ *
+ * Helpers for validation tests.
+ */
+final class ValidationRequestMocking {
```
Would this be maybe better style as a trait so static methods aren't required?

johnwatkins0 replied:
Done in 2e484276cea28486f5366917ee7a989b84a26958

westonruter commented on tests/php/test-class-amp-cli-validation-command.php:
```diff
@@ -6,17 +6,18 @@
+ *
+ * @coversDefaultClass AMP_CLI_Validation_Command
  */
 class Test_AMP_CLI_Validation_Command extends \WP_UnitTestCase {
```
```suggestion
class Test_AMP_CLI_Validation_Command extends WP_UnitTestCase {
```

westonruter commented:
(Sorry for the delay in reviewing.)

johnwatkins0 committed:
Combine redundant calls to get_urls method

johnwatkins0 committed:
Fix for line inadvertently placed in conditional

johnwatkins0 committed:
Formatting fixes

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Eliminate the use of a flag in validate URL function

johnwatkins0 committed:
Add leading slash to unnamespaced class used within namespace

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Convert ValidationRequestMocking into trait

johnwatkins0 committed:
Merge branch 'feature/refactor-url-validation' of https://github.com/ampproject/amp-wp into feature/refactor-url-validation

johnwatkins0 committed:
Reset lock timeout from inside CLI script

johnwatkins0 committed:
Eliminate lock timeout filter

johnwatkins0 committed:
PHPCS fix
